[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states

fegin opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #103784
* #103783
* #103782
* #103781
* __->__ #103609

Users may prefix the keys optim state_dict. Using`optim.state_dict()` to get the initial states is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

fegin committed:
[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states

Users may prefix the keys of `optim.state_dict()` is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_optim_utils.py:
nit: Shall we preserve the comment that we are specially keeping dim-0 states like `"step"` on CPU? (same with above in `_broadcast_processed_state()`)

awgu commented on torch/distributed/fsdp/_optim_utils.py:
Is there any concern that there may be a mixture of CPU and GPU states? (I am guessing that `t.device` might be GPU?)

awgu commented on torch/distributed/fsdp/_optim_utils.py:
nit: Should we add an assert message here just in case? For example, it may be helpful to at least include `"{len(fqns)} {fqns}` (so we know whether it was `len(fqns) !=` or `not use_orig_params` and have some sense of the FQNs).

awgu reviewed (approved):
Thanks for the fix!

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


empty states

Users may prefix the keys of `optim.state_dict()` is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin commented on torch/distributed/fsdp/_optim_utils.py:
I don't think this is an issue as the step is generally on CPU and other states may be moved to GPU. Moreover, this one is specific to the zero-size states which are not going to be used because the parameters are empty.

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


empty states

Users may prefix the keys of `optim.state_dict()` is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin commented on torch/distributed/fsdp/_optim_utils.py:
I don't think this is an issue as the step is generally on CPU and other states may be moved to GPU. Moreover, this one is specific to the zero-size states which are not going to be used because the parameters are empty.

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


empty states

Users may prefix the keys of `optim.state_dict()` is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


empty states

Users may prefix the keys of `optim.state_dict()` is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


empty states

Users may prefix the keys of `optim.state_dict()` is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


Users may prefix the keys of `optim.state_dict()` is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


Users may prefix the keys of `optim.state_dict()` is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


Users may prefix the keys optim state_dict. Using`optim.state_dict()` to get the initial states is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin committed:
Update on "[FSDP][optim_state_dict] Avoid calling optim.state_dict() to get the initial
empty states"


Users may prefix the keys optim state_dict. Using`optim.state_dict()` to get the initial states is brittle. This PR removes the call to `optim.state_dict()` and directly infers the empty states from the input states.

Differential Revision: [D46729119](https://our.internmc.facebook.com/intern/diff/D46729119/)

[ghstack-poisoned]

fegin commented:
The failing test is not related.
```
Usage:   
  pip install [options] <requirement specifier> [package-index-options] ...
  pip install [options] -r <requirements file> [package-index-options] ...
  pip install [options] [-e] <vcs project url> ...
  pip install [options] [-e] <local project path> ...
  pip install [options] <archive url/path> ...

no such option: --install-option
Error: Process completed with exit code 2.
```

fegin commented:
@pytorchbot merge -f "The failing test is not related"
